sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSink.java

```diff
@@ -0,0 +1,359 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.smb;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.extensions.smb.SMBFilenamePolicy.FileAssignment;
+import org.apache.beam.sdk.extensions.smb.SortedBucketFile.Writer;
+import org.apache.beam.sdk.extensions.smb.SortedBucketSink.WriteResult;
+import org.apache.beam.sdk.extensions.sorter.BufferedExternalSorter;
+import org.apache.beam.sdk.extensions.sorter.SortValues;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.schemas.transforms.Group;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Supplier;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Writes a PCollection representing sorted, bucketized data to files, where the # of files is equal
+ * to the # of buckets, assuming the contents of each bucket fit on a single worker.
+ *
+ * @param <SortingKeyT>
+ * @param <ValueT>
+ */
+public class SortedBucketSink<SortingKeyT, ValueT>
+    extends PTransform<PCollection<ValueT>, WriteResult> {
+
+  private final BucketMetadata<SortingKeyT, ValueT> bucketingMetadata;
+  private final SMBFilenamePolicy smbFilenamePolicy;
+  private final Supplier<Writer<ValueT>> writerSupplier;
+  private final ResourceId tempDirectory;
+  private final Coder<SortingKeyT> sortingKeyCoder;
+
+  public SortedBucketSink(
+      BucketMetadata<SortingKeyT, ValueT> bucketingMetadata,
+      Coder<SortingKeyT> sortingKeyCoder,
```

Comment:
I'm wondering if we disallow user specified key coder, but only infer it from `keyClass`, since using different coders in source/sink can mess up sorting?